eva_submission/assembly_taxonomy_insertion.py

```diff
@@ -0,0 +1,183 @@
+from lxml import etree
+import requests
+
+from ebi_eva_common_pyutils.pg_utils import get_all_results_for_query
+from ebi_eva_common_pyutils.logger import logging_config as log_cfg
+
+
+logger = log_cfg.get_logger(__name__)
+
+
+def download_xml_from_ena(accession):
+    """Download and parse XML from ENA using any accession or id (e.g. assembly or taxonomy)"""
+    try:  # catches any kind of request error, including non-20X status code
+        response = requests.get(f'https://www.ebi.ac.uk/ena/browser/api/xml/{accession}')
+        response.raise_for_status()
+    except requests.exceptions.RequestException as e:
+        raise e
+    root = etree.XML(bytes(response.text, encoding='utf-8'))
+    return root
+
+
+def get_assembly_name_and_taxonomy_id(assembly_accession):
+    xml_root = download_xml_from_ena(assembly_accession)
+    xml_assembly = xml_root.xpath('/ASSEMBLY_SET/ASSEMBLY')
+    if len(xml_assembly) == 0:
+        raise Exception(f'Assembly {assembly_accession} not found in ENA')
+    assembly_name = xml_assembly[0].get('alias')
+    taxonomy_id = int(xml_assembly[0].xpath('TAXON/TAXON_ID')[0].text)
+    return assembly_name, taxonomy_id
+
+
+def get_scientific_name_and_common_name(taxonomy_id):
+    xml_root = download_xml_from_ena(taxonomy_id)
+    xml_taxon = xml_root.xpath('/TAXON_SET/taxon')
+    if len(xml_taxon) == 0:
+        raise Exception(f'Taxonomy {taxonomy_id} not found in ENA')
+    scientific_name = xml_taxon[0].get('scientificName')
+    optional_common_name = xml_taxon[0].get('commonName')
+    return scientific_name, optional_common_name
+
+
+def get_assembly_set(conn, taxonomy, assembly_accession):
+    rows = get_all_results_for_query(conn,
+                         """SELECT acc.assembly_set_id 
+                         FROM evapro.accessioned_assembly acc 
+                         JOIN assembly_set asm on acc.assembly_set_id = asm.assembly_set_id 
+                         WHERE assembly_accession=\'{}\' AND taxonomy_id={}""".format(assembly_accession, taxonomy))
+
+    if len(rows) == 1:
+        return rows[0][0]
+    elif len(rows) == 0:
+        return None
+    else:
+        raise Exception('Inconsistent database state: several assembly_set_ids for the same taxonomy ({}) and assembly '
+                        'accession ({}): {}'.format(taxonomy, assembly_accession, rows))
+
+
+def is_taxonomy_in_evapro(conn, taxonomy_id):
```

Comment:
I know you're porting the code but this would be better as the method of a class. You could cache it and you would not have to pass the parameters around so much. Maybe in another refactor ...

Reply:
Agreed, as you can probably tell I took the strategy of least resistance here...